rclcpp_lifecycle/test/test_lifecycle_node_errors.cpp

```diff
@@ -0,0 +1,69 @@
+// Copyright 2024 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+#include "rcl_lifecycle/rcl_lifecycle.h"
+
+#include "rclcpp_lifecycle/lifecycle_node.hpp"
+
+#include "./mocking_utils/patch.hpp"
+
+class TestDefaultStateMachine : public ::testing::Test
+{
+protected:
+  static void SetUpTestCase()
+  {
+    rclcpp::init(0, nullptr);
+  }
+  static void TearDownTestCase()
+  {
+    rclcpp::shutdown();
+  }
+};
+
+class EmptyLifecycleNode : public rclcpp_lifecycle::LifecycleNode
+{
+public:
+  explicit EmptyLifecycleNode(const std::string & node_name)
+  : rclcpp_lifecycle::LifecycleNode(node_name)
+  {}
+};
+
+// This test is split out of test_lifecycle_node.cpp for an esoteric reason.  When running on
+// RedHat-based distributions (like Fedora or RHEL), the way that glibc is compiled does not
+// allow mocking_utils::inject_on_return to work.  Thus the test has to patch_and_return().
+// Unfortunately, this means that the resources are not actually cleaned up, and thus other tests
+// may return incorrect results.  By having it in a separate process we ensure that the resources
+// will at least be cleaned up by the process dying.
+TEST_F(TestDefaultStateMachine, empty_initializer_rcl_errors)
+{
+  {
+    auto patch = mocking_utils::patch_and_return(
+      "lib:rclcpp_lifecycle", rcl_lifecycle_state_machine_init, RCL_RET_ERROR);
+    EXPECT_THROW(
+      std::make_shared<EmptyLifecycleNode>("testnode").reset(),
```

Comment:
> I know this is an existing issue with the test and out of scope for this pr, but technically this is testing both the constructor and destructor

Since we are patching `rcl_lifecycle_state_machine_init` to fail, it means that the constructor of `LifecycleNodeInterfaceImpl` throws an exception.  In turn, that means that the constructor of `LifecycleNode` fails, and finally that means that the constructor of `EmptyLifecycleNode` fails.  Because of that chain of events, as long as things are working as expected the destructor will never run.

I guess in the case where the chain of constructors does *not* fail, then you are correct in that the destructor could then run, could throw an exception, and thus erroneously pass the test.  But I don't see a good way to fix this; in the nominal case we don't get an object back (because the constructor has failed).  I guess we could attempt to assign the `EmptyLifecycleNode` to an object, and if that succeeds then we know that this test failed?  Not sure it is worth it.

> Also the `.reset()` is completely redundant here.

Yeah, that is definitely true, though not a big deal, I guess.

I'm going to go ahead with this PR as-is, because I'm honestly not sure how to solve the other problems.  But I'm happy to make another change if you have an idea of how to improve this.